File: home_page.py
from __future__ import annotations

from PyQt6 import uic
from PyQt6.QtCore import pyqtSignal
from PyQt6.QtWidgets import QWidget

from telemetry_data import TelemetryData


class HomePage(QWidget):
    requestNavigation = pyqtSignal()
    requestCamera = pyqtSignal()

    def __init__(self, parent: QWidget | None = None) -> None:
        super().__init__(parent)
        uic.loadUi("homepage.ui", self)
        self._telemetry: TelemetryData | None = None

        self.btnNav.clicked.connect(self.requestNavigation)
        self.btnCam.clicked.connect(self.requestCamera)

    def bind_telemetry(self, telemetry: TelemetryData) -> None:
        self._telemetry = telemetry
        if self._telemetry is None:
            return

        refresh = self._refresh
        refresh()

        self._telemetry.speedKmhChanged.connect(refresh)
        self._telemetry.batteryPercentChanged.connect(refresh)
        self._telemetry.gpsOkChanged.connect(refresh)
        self._telemetry.reverseChanged.connect(refresh)

    def _refresh(self) -> None:
        if self._telemetry is None:
            return

        self.lblSpeedBig.setText(str(int(round(self._telemetry.speedKmh))))
        self.lblBattery.setText(f"{self._telemetry.batteryPercent}%")
        self.lblGps.setText("GPS OK" if self._telemetry.gpsOk else "GPS LOST")
        self.lblReverse.setText("R" if self._telemetry.reverse else "D")

File: homepage.cpp
#include "homepage.h"
#include "ui_homepage.h"
#include "telemetrydata.h"

HomePage::HomePage(QWidget* parent) : QWidget(parent), ui(new Ui::HomePage) {
    ui->setupUi(this);
    connect(ui->btnNav, &QPushButton::clicked, this, &HomePage::requestNavigation);
    connect(ui->btnCam, &QPushButton::clicked, this, &HomePage::requestCamera);
}
HomePage::~HomePage(){ delete ui; }

void HomePage::bindTelemetry(TelemetryData* t){
    m_t=t;
    if(!m_t) return;

    auto refresh = [this](){
        ui->lblSpeedBig->setText(QString::number((int)qRound(m_t->speedKmh())));
        ui->lblBattery->setText(QString("%1%").arg(m_t->batteryPercent()));
        ui->lblGps->setText(m_t->gpsOk() ? "GPS OK" : "GPS LOST");
        ui->lblReverse->setText(m_t->reverse() ? "R" : "D");
    };
    refresh();

    connect(m_t, &TelemetryData::speedKmhChanged, this, refresh);
    connect(m_t, &TelemetryData::batteryPercentChanged, this, refresh);
    connect(m_t, &TelemetryData::gpsOkChanged, this, refresh);
    connect(m_t, &TelemetryData::reverseChanged, this, refresh);
}
